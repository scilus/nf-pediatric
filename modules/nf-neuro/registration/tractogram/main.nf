process REGISTRATION_TRACTOGRAM {
    tag "$meta.id"
    label 'process_dynamic'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(anat), path(transfo), path(tractogram), path(ref) /* optional, value = [] */, path(deformation) /* optional, value = [] */

    output:
    tuple val(meta), path("*__*.{trk,tck,h5}"), emit: warped_tractogram
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ? "_${task.ext.suffix}" : ""
    def reference = "$ref" ? "--reference $ref" : ""
    def in_deformation = "$deformation" ? "--in_deformation $deformation" : ""

    def inverse = task.ext.inverse ? "--inverse" : ""
    def reverse_operation = task.ext.reverse_operation ? "--reverse_operation" : ""
    def force = task.ext.force ? "-f" : ""

    def cut_invalid = task.ext.cut_invalid ? "--cut_invalid" : ""
    def remove_single_point = task.ext.remove_single_point ? "--remove_single_point" : ""
    def remove_overlapping_points = task.ext.remove_overlapping_points ? "--remove_overlapping_points" : ""
    def threshold = task.ext.threshold ? "--threshold " + task.ext.threshold : ""
    def no_empty = task.ext.no_empty ? "--no_empty" : ""

    """
    for tractogram in ${tractogram};
        do \
        ext=\${tractogram#*.}
        bname=\$(basename \${tractogram} .\${ext})

        if [[ \$ext == "h5" ]]; then
            scil_tractogram_apply_transform_to_hdf5.py \$tractogram $anat $transfo \
                        ${prefix}__\${bname}${suffix}.\${ext} \
                        $in_deformation\
                        $inverse\
                        $cut_invalid\
                        $reverse_operation\
                        $force\
                        $reference
        else
            scil_tractogram_apply_transform.py \$tractogram $anat $transfo tmp.trk\
                            $in_deformation\
                            $inverse\
                            $cut_invalid\
                            $reverse_operation\
                            $force\
                            $reference

            scil_tractogram_remove_invalid.py tmp.trk ${prefix}__\${bname}${suffix}.\${ext}\
                            $cut_invalid\
                            $remove_single_point\
                            $remove_overlapping_points\
                            $threshold\
                            $no_empty\
                            -f
        fi
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ? "_${task.ext.suffix}" : ""
    """
    scil_tractogram_apply_transform.py -h
    scil_tractogram_remove_invalid.py -h

    for tractogram in ${tractogram};
        do \
        ext=\${tractogram#*.}
        bname=\$(basename \${tractogram} .\${ext})

        touch ${prefix}__\${bname}${suffix}.\${ext}
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """
}
