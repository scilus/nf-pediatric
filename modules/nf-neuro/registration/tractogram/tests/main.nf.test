
nextflow_process {

    name "Test Process REGISTRATION_TRACTOGRAM"
    script "../main.nf"
    process "REGISTRATION_TRACTOGRAM"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/tractogram"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "bundles.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }


    test("registration - tractogram_bundles") {


        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/bundle_all_1mm.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/affine.txt", checkIfExists: true),
                        file("\${test_data_directory}/fibercup_atlas/subj_1/*.trk", checkIfExists: true),
                        [],
                        []
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("registration - tractogram") {

        config "./nextflow_suffix.config"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/bundle_all_1mm.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/affine.txt", checkIfExists: true),
                        file("\${test_data_directory}/fibercup_atlas/subj_1/bundle_2.trk", checkIfExists: true),
                        [],
                        []
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }
}
