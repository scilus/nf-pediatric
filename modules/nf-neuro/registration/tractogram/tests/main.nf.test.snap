{
    "registration - tractogram_bundles": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_0.trk:md5,6e14cc02b66d12d5dde0a0701918385a",
                            "test__bundle_1.trk:md5,ef5a759144ab7e8d2f5abf30a463a152",
                            "test__bundle_2.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9",
                            "test__bundle_3.trk:md5,8877161b9aebaa6b12a4d7a8f8410407",
                            "test__bundle_4.trk:md5,6c8658b946154896952ca89e88fcf29f",
                            "test__bundle_5.trk:md5,fa058f93d9bfe4bed2ff027e9acd5682",
                            "test__bundle_6.trk:md5,125e50561b4b0c5fd8ee23867c28bc51"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,609469c08e60062eed7528276deddf5e"
                ],
                "versions": [
                    "versions.yml:md5,609469c08e60062eed7528276deddf5e"
                ],
                "warped_tractogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_0.trk:md5,6e14cc02b66d12d5dde0a0701918385a",
                            "test__bundle_1.trk:md5,ef5a759144ab7e8d2f5abf30a463a152",
                            "test__bundle_2.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9",
                            "test__bundle_3.trk:md5,8877161b9aebaa6b12a4d7a8f8410407",
                            "test__bundle_4.trk:md5,6c8658b946154896952ca89e88fcf29f",
                            "test__bundle_5.trk:md5,fa058f93d9bfe4bed2ff027e9acd5682",
                            "test__bundle_6.trk:md5,125e50561b4b0c5fd8ee23867c28bc51"
                        ]
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-15T19:11:42.198039"
    },
    "registration - tractogram": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundle_2_mni.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,609469c08e60062eed7528276deddf5e"
                ],
                "versions": [
                    "versions.yml:md5,609469c08e60062eed7528276deddf5e"
                ],
                "warped_tractogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundle_2_mni.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-15T19:11:52.03634"
    }
}